### PR TITLE
Feature/issue 311 newton solver

### PR DESCRIPTION
**Pull request by charlesm93**
## Summary
Solve issue#311 and expose the Newton solver from Kinsol to the Stan language. The name of the new function is:
* `algebra_solver_newton()

It has the exact same signature as the original `algebra_solver`.

## Details
This is the first step to exposing the Newton solver. The next step is to deprecate `algebra_solver`, replace the current version with `algebra_solver_powell`, and work on variadic arguments (as currently done with the ODE integrator); and potentially on closures too.

## Tests:
* `test/integration/good/algebra_solver_newton.stan`
* `test/integration/bad/bad_newton_*.stan`

Copyright holder: Charles Margossian, Columbia University

The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
* Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
* Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by charlesm93** [comment]
@rok-cesnovar or @SteveBronder would you mind reviewing this? It should be straightforward.

**Comment by charlesm93** [comment]
I do owe everyone an apology for taking this long to do this. This should have been in a year ago. I'll be more eager to bite the bullet from now on.

File: src/stan_math_backend/Expression_gen.ml [hunk 1]
```diff
@@ -297,6 +297,8 @@ and gen_fun_app ppf fname es =
       match (is_hof_call, fname, converted_es @ extra) with
       | true, "algebra_solver", f :: x :: y :: dat :: datint :: tl ->
           (fname, f :: x :: y :: dat :: datint :: msgs :: tl)
+      | true, "algebra_solver_newton", f :: x :: y :: dat :: datint :: tl ->
+          (fname, f :: x :: y :: dat :: datint :: msgs :: tl)
```
Comment (rok-cesnovar) [inline src/stan_math_backend/Expression_gen.ml:301]:
```suggestion
      | true, "algebra_solver", f :: x :: y :: dat :: datint :: tl
      | true, "algebra_solver_newton", f :: x :: y :: dat :: datint :: tl ->
          (fname, f :: x :: y :: dat :: datint :: msgs :: tl)
```
When 2 patterns have the same code, you can shorten it a bit.

**Comment by charlesm93** [comment]
@rok-cesnovar Added the changes you suggested. The code should be ready for a final review.

**Pull request merged**